### ISSUE TITLE
Working with many variables can be slow

### ISSUE THREAD

**Issue by matiasdahl** [sympy/sympy#10232]
It seems that sympy can be very slow when expressions contain a lot of variables. The below code illustrate this by computing an inner product of one numeric vector of length 1500 and a symbolic vector with 1500 variables. The inner product is computed in two different ways and both approaches take around 100 seconds on a 3GHz/8GB system. The second approach shows that adding terms is fast in the beginning, but towards the end of sum, each additional term takes more and more time.

sympy: 0.7.6.1. python 3.4. 

```
import sympy as sp

# Create a symbolic 1x1500 row vector with variables a0, ..., a1499
v = sp.Matrix(sp.symbols('a0:1500')).T

# Create a 1500x1 column vector with some numerical values
x = sp.Matrix([k%7 - k%5 for k in range(1500)])

# Approach 1: Compute inner product of matrices. (c. 1:45 minutes)
result1 = (v * x)[0, 0]

# Approach 2: Compute inner product manually (also c. 1:45 minutes)
result2 = 0
for k in range(1500):
    if k % 10 == 0:
        print(k)
    result2 += v[k] * x[k]

# Verify that both approaches give the same result
assert(sp.simplify(result1 - result2) == 0)
```

**Comment by jksuom** [comment]
It is possible to choose a quicker way:
``>>> result = sp.Add(*[v[k]*x[k] for k in range(1500)])``

**Comment by matiasdahl** [comment]
Thank you! That is much better. The runtime for the sum drops from c. 100 sec to 1 s. I experimented a bit, and if someone else is also trying to speed up these kinds of inner products, I noticed that the runtime also depends on the number of variables that are currently defined.

The below will run in 1.2 secs:
```
v = sp.Matrix(sp.symbols('a0:1500')).T
x = sp.Matrix([k%7 - k%5 for k in range(1500)])
result = sp.Add(*[v[k]*x[k] for k in range(1500)])
```

However, if one adds the below line before the code, the runtime will be around 10 seconds.
```
extra_variables_not_used = sp.symbols('e0:200000')
```
